File: cplusplus/k_closest_points.cpp
//
//  k_closest_points.cpp
//  LintCode
//
//  Created by 张夜白 on 2017/11/30.
//  Copyright © 2017年 张夜白. All rights reserved.
//

#include <stdio.h>
#include <vector>
#include <algorithm>
#include <queue>
#include <iostream>
using namespace std;

struct Point {
    int x;
    int y;
    Point() : x(0), y(0) {}
    Point(int a, int b) : x(a), y(b) {}
};

static int distanceOfTwoPoints(const Point & src, const Point & dest);
static int square(int x);
static int distanceOfTwoPoints(const Point & src, const Point & dest){
    return square(src.x - dest.x) + square(src.y - dest.y);
}
static int square(int x){
    return x * x;
}
auto global_point = Point(0,0);
auto  compare = [](const Point & l_point, const Point & r_point){
    // lpd: left point distance between itself to global_point
    int lpd = distanceOfTwoPoints(l_point, global_point);
    int rpd =distanceOfTwoPoints(r_point, global_point);
    if(lpd == rpd){
        if(l_point.x == r_point.x)
            return l_point.y < r_point.y;
        else
            return l_point.x < r_point.x;
    }else{
        return lpd < rpd;
    }
};
class K_Closest_Points {
public:
    /*
     * @param points: a list of points
     * @param origin: a point
     * @param k: An integer
     * @return: the k closest points
     */
    vector<Point> kClosest(vector<Point> points, Point origin, int k) {
        // write your code here
        global_point = origin;
        std::priority_queue<Point, std::vector<Point>, decltype(compare)> p_queue(compare);
        vector<Point> result;
        for(const auto & point: points){
            p_queue.push(point);
            if(p_queue.size() > k)
                p_queue.pop();

        }
        for(; !p_queue.empty();){
            const Point & key = p_queue.top();
            result.push_back(key);
            p_queue.pop();
        }
        std::reverse(result.begin(), result.end());
        return result;
    }
};

int main(int argc, char * argv[]){
    std::vector<Point> points = {Point(4,6),Point(4,7),Point(4,4),Point(2,5),Point(1,1),Point(1, 0)};
    Point origin = Point(0, 0);
    int countOfPoints = 3;
    K_Closest_Points k_closest_points;
    vector<Point>  result = k_closest_points.kClosest(points, origin, countOfPoints);
    std::cout << "[";
    for(const auto & point : result){
        std::cout << "[" << point.x << "," << point.y << "]";
    }
    std::cout << "]" << std::endl;
}

File: java/K_Closest_Points.java
import java.util.PriorityQueue;
import java.util.Arrays;
import java.util.Comparator;

public class K_Closest_Points
{
	public Point [] kClosest(Point[] points, Point origin, int k){
		Point[] result  = new Point[k];
		PriorityQueue<Point> p_queue = new PriorityQueue<Point>( new Comparator<Point>(){
			private int distanceOfTwoPoints(Point l_p, Point r_p){
				return this.square(l_p.x - r_p.x) + this.square(l_p.y - r_p.y);
			}

			private int square(int src){
				return src * src;
			}

			public int compare(Point l_p, Point r_p){
				int l_distance = this.distanceOfTwoPoints(l_p, origin);
				int r_distance = this.distanceOfTwoPoints(r_p, origin);
				if(l_distance == r_distance){
					if(r_p.x == l_p.x){
						return r_p.y - l_p.y;
					}
					else{
						return r_p.x - l_p.x;
					}
				}
				else{
					return r_distance - l_distance;
				}
			}
		});
		for(Point point : points){
			p_queue.add(point);
			if(p_queue.size() > k){
				p_queue.poll();
			}
		}
		//Object [] tempArr = p_queue.toArray();  返回的是无序数组，并不一定是无序的，总之顺序不保证。
		for(int i = k - 1; (!p_queue.isEmpty()) && (i >= 0); --i){
			result[i] = p_queue.poll();
		}
		return result;
	}
	public static void main(String [] args){
		Point [] points = {new Point(4,6),new Point(4,7),new Point(4,4),new Point(2,5),new Point(1,1),new Point(1, 0)};
		Point origin = new Point(0, 0);
		int k = 3;
		Point [] result = new K_Closest_Points().kClosest(points, origin, k);
		System.out.print("[");
		for(Point point : result){
			System.out.print("[" + point.x + "," + point.y + "]");
		}
		System.out.println("]");
		
		int [] S = {3,4,6,7};
		System.out.print(new K_Closest_Points().triangleCount(S));
	}

	public int triangleCount(int[] S) {
		// write your code here
		int[] tempArray = S;
		Arrays.sort(tempArray);
		int size = S.length;
		int result = 0;
		//boolean out = false;
		for(int index = 0; index < size - 2; ++index){
			//int j = index + 1;
			out = false;
			for(int j = size - 1; j > index + 1; --j){
				if(tempArray[j] - tempArray[j - 1] < tempArray[index]){
					int k = index + 1;
					for(; k < j; ++k){
						if(tempArray[j] < tempArray[k] + tempArray[index]){
							result += j - k;
							break;
						}
					}
				}
			}
		}
        return result;
    }

}
class Point
{
	int x;
	int y;
	Point(){
		this.x = 0;
		this.y = 0;
	}
	Point(int a, int b){
		this.x = a;
		this.y = b;
	}
}